### PR TITLE
Refactor CMP and CMPState

### PR DESCRIPTION
**Pull request by merajhashemi**
### CMP and CMPState
- Converted CMPState into a dataclass for improved readability and maintainability.
- Removed dual attributes from `LagrangianStore`. 
- Moved the `backward` method from CMPState to `LagrangianStore`.
- Utilized LagrangianStore as a data structure to organize information within `CMPState` replacing direct attributes such as `self.lagrangian` and `self.dual_lagrangian`.
- Added constraints attribute and related helper methods to CMP
- Renamed populate_*_lagrangian to compute_*_lagrangian method
- Removed purge_* methods for lagrangians
- Added `dual_parameters()` method for fetching multiplier parameters for instantiating the dual optimizer

### PenaltyCoefficients
- Transformed PenaltyCoefficients to no longer be `nn.module`.
- Wrapped `__call__` in `no_grad` mode for efficiency.
- Implemented `def __call__(self, *args, **kwargs):` in the base class for consistency.
- Adjusted `__call__` to return cloned values.
- Moved penalty coefficient update parameters out of Formulation and into PenaltyCoefficientUpdater.
- Introduced `PenaltyCoefficientUpdater` to handle update schemes for penalty coefficients.
- Implemented Multipicative

Updates from #88
### Constraint
- Extracted `ConstraintType` enum to separate module to resolve circular dependency issues between `Constraint`, `Formulation` and `Multiplier`.

### Formulations
- Removed `PenaltyFormulation` and `QuadraticPenaltyFormulation`
  - `PenaltyFormulation` is deprecated in favor of asking users to formulate penalized tasks directly through their objective function
  - We retain some utility functions for the `QuadraticPenaltyFormulation`. We will add a tutorial to illustrate the use of a Quadratic  Penalty formulation in Cooper.
- Only `LagrangianFormulation` and `AugmentedLagrangianFormulation` remain. Now it is mandatory to provide a `Multiplier` when building a formulation.
- Removed `FormulationType`.
- Removed formulation kwargs such as `penalty_growth_factor` and `violation_tolerance`. 

### Multipliers
- Now expect `init` tensor (if provided) to have shape `(num_constraints,)`. Else, `num_constraints:int` is expected and we initialize the multipliers with zeros of shape `(num_constraints,)`.
- Removed support for `dual_restarts`.
- Refactored `enforce_positive` attribute to `is_inequality` property.
- Gathered equality and inequality multiplier tests into joint tests for `ExplicitMultipliers`

Updates from #89
- Changed observed_constraints attribute of CMPState to be a dictionary with Constraint keys and ConstraintState values
- Similarly, refactored constraint_measurements to be a dictionary with Constraint keys and ConstraintMeasurement values
- Added utility function for retrieving observed_multiplier_values from LagrangianStore
- Added utility function for retrieving observed_violations and observed_violations from CMPState

Updates from #90 
- Implement `CooperOptimizer` abstract base class
  - Both `ConstrainedOptimizer` and `UnconstrainedOptimizer` inherit from `CooperOptimizer`
- Add `RollOut` for organizing returns of `roll` method
- Remove `ConstraintMeasurement`
- Add `ContributionStore`
- Remove AugmentedLagrangian optimizers
- Fix Penalty Coefficient and Updater to allow for scalar values
- Remove Slacks

### CODE REVIEW

File: cooper/optim/utils.py [hunk 1]
```diff
@@ -2,9 +2,9 @@
 
 import torch
 
-from cooper.multipliers import Multiplier
 from cooper.utils import OneOrSequence, ensure_sequence
 
+from .. import ConstrainedMinimizationProblem
 from . import constrained_optimizers
 from .optimizer_state import CooperOptimizerState
 from .types import AlternationType
```
Comment (juan43ramirez) [inline cooper/optim/utils.py:10]: Absolute imports vs relative imports discussion: consistency

File: cooper/optim/utils.py [hunk 2]
```diff
@@ -13,8 +13,8 @@
 
 def create_optimizer_from_kwargs(
     primal_optimizers: OneOrSequence[torch.optim.Optimizer],
+    cmp: ConstrainedMinimizationProblem,
     dual_optimizers: Optional[OneOrSequence[torch.optim.Optimizer]] = None,
```
Comment (juan43ramirez) [inline cooper/optim/utils.py:17]: I forget, did we agree to keep this function? It is only used for the tests.

Should we add a TODO?

File: cooper/formulations/utils.py [hunk 3]
```diff
@@ -85,61 +75,62 @@ def compute_dual_weighted_violation(
...
-        if multiplier_value is None:
-            linear_contribution = 0.0
-        else:
-            linear_contribution = compute_primal_weighted_violation(
-                constraint_factor_value=multiplier_value, violation=violation
-            )
-        quadratic_penalty = 0.5 * torch.einsum("i...,i...->", penalty_coefficient_value, violation**2)
-        return linear_contribution + quadratic_penalty
+        # TODO(gallego-posada): Why is the linear term being computed with the "primal" weighted violation?
+        # Should we also have a "dual_weighted_violation"?
```
Comment (juan43ramirez) [inline cooper/formulations/utils.py:131]: Here we are computing the primal component of the augmented Lagrangian. It makes sense to use the primal weighted violation computation to (1) preserve gradients wrt the violations and (2) block gradients wrt the multipliers

File: cooper/optim/constrained_optimizers/alternating_optimizer.py [hunk 4]
```diff
@@ -89,9 +67,9 @@ class AlternatingPrimalDualOptimizer(BaseAlternatingOptimizer):
 
     def roll(
         self,
-        compute_cmp_state_fn: Callable[..., CMPState],
-        compute_violations_fn: Optional[Callable[..., CMPState]] = None,
-    ) -> tuple[CMPState, LagrangianStore]:
+        compute_cmp_state_kwargs: dict = {},
+        compute_violations_kwargs: dict = {},
+    ) -> tuple[CMPState, LagrangianStore, LagrangianStore]:
```
Comment (juan43ramirez) [inline cooper/optim/constrained_optimizers/alternating_optimizer.py:72]: Remember to update type hint after returning the loss.
Also, could define the type above so that the signature of roll fits in one/three lines

File: cooper/optim/constrained_optimizers/alternating_optimizer.py [hunk 5]
```diff
@@ -156,44 +134,24 @@ def roll(
...
-            if compute_violations_fn is not None:
-                new_cmp_state = compute_violations_fn()
+            try:
+                new_cmp_state = self.cmp.compute_violations(**compute_violations_kwargs)
 
                 if new_cmp_state.loss is not None:
                     raise RuntimeError(
-                        "Expected `compute_violations_fn` to not populate the loss. Please provide this value for the `compute_cmp_state_fn` instead."
+                        "Expected `compute_violations` to not populate the loss. "
+                        "Please provide this value for the `compute_cmp_state` instead."
```
Comment (juan43ramirez) [inline cooper/optim/constrained_optimizers/alternating_optimizer.py:143]: Why not use triple quotes to wrap all of the text?

File: cooper/optim/constrained_optimizers/extrapolation_optimizer.py [hunk 6]
```diff
@@ -58,13 +53,6 @@ def custom_sanity_checks(self):
...
             )
 
-        for multiplier in self.multipliers:
-            if getattr(multiplier, "restart_on_feasible", False):
-                raise RuntimeError(
-                    """Using restart on feasible for multipliers is not supported in
-                    conjunction with the ExtrapolationConstrainedOptimizer."""
-                )
-
     def step(self, call_extrapolation: bool = False):
```
Comment (juan43ramirez) [inline cooper/optim/constrained_optimizers/extrapolation_optimizer.py:56]: `@torch.no_grad`

File: cooper/penalty_coefficient_updaters/penalty_coefficient_updaters.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        violation, strict_violation = constraint_state.extract_violations()
+        constraint_features, strict_constraint_features = constraint_state.extract_constraint_features()
+        penalty_coefficient = constraint.penalty_coefficient
+
+        # TODO(merajhashemi): Make this block more readable
+        values_for_observed = (
+            penalty_coefficient.value
+            if isinstance(penalty_coefficient, DensePenaltyCoefficient)
+            else penalty_coefficient.value[strict_constraint_features]
+        )
```
Comment (juan43ramirez) [inline cooper/penalty_coefficient_updaters/penalty_coefficient_updaters.py:59]: Could you use the `extract_*` functions we use for the formulations?

File: cooper/cmp.py [hunk 8]
```diff
@@ -1,42 +1,44 @@
...
-    dual_constraint_measurements: Optional[list[ConstraintMeasurement]] = None
+    lagrangian: Optional[torch.Tensor] = None
+    measured_constraints: dict[Constraint, ConstraintMeasurement] = field(default_factory=dict)
 
-    def multiplier_values_for_primal_constraints(self):
-        return [_.multiplier_value for _ in self.primal_constraint_measurements]
+    def backward(self) -> None:
+        """Triggers backward calls to compute the gradient of the Lagrangian with
+        respect to the primal variables."""
+        if self.lagrangian is not None and isinstance(self.lagrangian, torch.Tensor):
```
Comment (juan43ramirez) [inline cooper/cmp.py:24]: Why would the Lagrangian not be a Tensor?
This check sounds more like an `assert isinstante(...)` to me

File: cooper/cmp.py [hunk 9]
```diff
@@ -1,42 +1,44 @@
...
-    def multiplier_values_for_dual_constraints(self):
-        return [_.multiplier_value for _ in self.dual_constraint_measurements]
+    def observed_multiplier_values(self):
+        for constraint_state in self.measured_constraints.values():
+            yield constraint_state.multiplier_value
 
 
+@dataclass
 class CMPState:
+    # TODO(gallego-posada): consider adding utilities for fetching constraint features
```
Comment (juan43ramirez) [inline cooper/cmp.py:34]: sure, why not

**Review by juan43ramirez — CHANGES_REQUESTED** [review]
Bug in AugmentedLagrangian.
Otherwise, minor comments here and there

File: cooper/optim/constrained_optimizers/extrapolation_optimizer.py (hunk 6 above)
Comment (gallego-posada) [reply cooper/optim/constrained_optimizers/extrapolation_optimizer.py:56]: You mean for decorating `step`? 
I don't think that is correct for extrapolation. Remember that there is an extra ;) gradient computation inside this step, so you can't wrap this method in `no_grad`

File: cooper/optim/constrained_optimizers/alternating_optimizer.py (hunk 4 above)
Comment (gallego-posada) [reply cooper/optim/constrained_optimizers/alternating_optimizer.py:72]: This type idea has been captured by the `RollOut` dataclass implemented in #90

File: cooper/formulations/formulations.py [hunk 10]
```diff
@@ -256,21 +153,19 @@ def compute_contribution_for_dual_lagrangian(
...
+        dual_lagrangian_contribution = formulation_utils.compute_dual_weighted_violation(
             constraint_factor_value=multiplier_value,
             violation=strict_violation,
             penalty_coefficient_value=penalty_coefficient_value,
         )
-        dual_constraint_store = ConstraintMeasurement(
-            violation=strict_violation,
-            multiplier_value=multiplier_value,
+        dual_constraint_measurement = ConstraintMeasurement(
+            violation=strict_violation, multiplier_value=multiplier_value
```
[the remark this answers is not in the document]
Comment (gallego-posada) [reply cooper/formulations/formulations.py:159]: This is now being handled by `ContributionStore` implemented in #90

File: cooper/formulations/utils.py (hunk 3 above)
Comment (juan43ramirez) [reply cooper/formulations/utils.py:131]: Consider renaming this method to indicate that it only deals with _primal_ contributions.
This would address the TODO

File: cooper/formulations/utils.py [hunk 11]
```diff
@@ -85,61 +75,62 @@ def compute_dual_weighted_violation(
...
-
         aux1 = torch.einsum("i...,i...->i...", penalty_coefficient_value, violation)
-
-        if multiplier_value is None:
-            return 0.5 * torch.einsum("i...,i...->", 1 / penalty_coefficient_value, torch.relu(aux1) ** 2)
-        else:
-            aux2 = torch.relu(multiplier_value + aux1) ** 2 - multiplier_value**2
-            return 0.5 * torch.einsum("i...,i...->", 1 / penalty_coefficient_value, aux2)
-
+        aux2 = torch.relu(multiplier_value + aux1) ** 2 - multiplier_value**2
```
[the remark this answers is not in the document]
Comment (juan43ramirez) [reply cooper/formulations/utils.py:127]: Confirmed together that  `multiplier_value` should be detached.

File: cooper/cmp.py (hunk 8 above)
Comment (merajhashemi) [reply cooper/cmp.py:24]: resolved in #90

File: cooper/penalty_coefficient_updaters/penalty_coefficient_updaters.py (hunk 7 above)
Comment (merajhashemi) [reply cooper/penalty_coefficient_updaters/penalty_coefficient_updaters.py:59]: Not really, the `extract_*` does not differentiate between indexed and dense penalty coefficients; it only checks if the `constraint_features` is None or not.

File: cooper/constraints/constraint_type.py [hunk 12]
```diff
@@ -0,0 +1,8 @@
+from enum import Enum, auto
+
+
+class ConstraintType(Enum):
+    """Constraint type enumeration."""
+
+    EQUALITY = auto()
+    INEQUALITY = auto()
```
[the remark this answers is not in the document]
Comment (merajhashemi) [reply cooper/constraints/constraint_type.py:8]: Added TODO

[the remark this answers is not in the document]
Comment (gallego-posada) [reply cooper/constraints/constraint_type.py:8]: Added TODO.

File: cooper/cmp.py [hunk 13]
```diff
@@ -29,6 +66,7 @@ def multiplier_values_for_dual_constraints(self):
         return [_.multiplier_value for _ in self.dual_constraint_measurements]
 
 
+@dataclass
 class CMPState:
```
[the remark this answers is not in the document]
Comment (gallego-posada) [reply cooper/cmp.py:36]: Methods reordered consistently

**Review by juan43ramirez — APPROVED** [review]
